Add and edit Storage DG (#129)
AY2425S1-CS2113-W14-1/tp: docs/DeveloperGuide.md

spending. This information was used by other classes to perform their component tasks.

### Storage Component
#### Motivation behind the component:
+ Allows the user to save changes, so that they can resume where they left off.
+ Allows advanced users to edit files directly, enabling fast, manual adjustments.

#### How the Storage Component works:
+ Variables and File Path:
    + `incomes` → `./incomes.txt`
    + `spendings` → `./spendings.txt`
    + `password` → `./password.txt`

+ To save edited lists:

>     + It is done when users type `bye`, which signals the end of the program.

Reviewer: ```suggestion
    + It is done when users type `bye` or after keyboard interrupts (ie.Ctrl-c), which signals the end of the program.
```

>     + The lists are serialised to a user-editable format in their respective files.

Reviewer: Not sure if serialised is still the right word

+ To load saved lists:
  + It is done upon program startup, when `Wiagi` is constructed.
  + Within `Wiagi` constructor, it will create a new instance of `Storage`, which will then de-serialise the data at the 
  `incomes` and `spendings` file paths to an `IncomeList` and `SpendingList` respectively.
  + `Wiagi` will then retrieve the lists in `Storage` to initialise its lists.
  + Data corruption in the file triggers an exception, often due to user-editing errors.
  + For missing files (e.g., new users), files are created and the initialised lists will be empty..

+ To load password:
  + The hashed password will simply be loaded from the password file.
  + For missing files (e.g., new users), users will be prompted to set a new password at the start of the
  program. The entered password will then be hashed and stored in a newly created password file.

#### Implementation:
#### Storage class
The `Storage` class is a class that stores `incomes`, `spendings` and `password`. 
Upon instantiation, it will call `IncomeListStorage.load()`, `SpendingListStorage.load()` and `LoginStorage.load()`, 
which will initialise the variables in `Storage` respectively.

> #### save method in `IncomeListStorage` `SpendingListStorage`

Reviewer: ```suggestion
#### Save method in `IncomeListStorage` `SpendingListStorage`
```

> Both classes have similar implementation for `save()`, except that `SpendingListStorage` saves budget details on the 
> first line.

Reviewer: ```suggestion
Both classes have similar implementation for `save()`, except that `SpendingListStorage` saves budget details in the 
first line of its respective text file.
```

+ Format: `daily budget | monthly budget | yearly budget`
+ A for loop will loop through the list, and get each of the attributes of each entry within it and separate them by 
`|`. Hence, each entry will be written line by line to the file.
+ Format: `amount | description | date | tag | recurrence frequency | last recurrence date | last recurrence day`
  + E.g. `add income 10 part time /2024-10-10/ *job* ~monthly~` will be stored as
    `10.0|part time|2024-10-10|job|MONTHLY|2024-10-10|10`

> ![storageSave.png](./Diagrams/Storage/saveStorageSequenceDiagram.drawio.png)

Reviewer: For some reason the diagram are inverted in color, cant see much.

Reviewer: same for the other storage diagrams

Author: I have updated the diagrams, please let me know if the problem still persists.

#### load method in `IncomeListStorage` `SpendingListStorage`
Both classes have similar implementation for `load()`, except that `SpendingListStorage` also loads budget details.
+ A while loop will loop through the file with a scanner to read line by line till the end of the file is reached.
+ It splits each line by `|` to access each attributes, convert date and last recurrence date to `LocalDate` type, 
and add it to the lists.
+ During the process, if a line is corrupted, an exception will be caught and user will be informed.
![storageLoad.png](./Diagrams/Storage/loadStorageSequenceDiagram.drawio.png)

#### load method in `LoginStorage`
+ It first checks if the password file exists.
  + If yes, it will use a scanner to read the file and initialise `password` in `Storage`.
  + Else, it will call `createNewUser()`, which creates a new password file and use `getNewUserPassword()` to scan for
  the user input. Then, it will be hashed, stored in the file, and be used to initialise `password` in `Storage`.

> ![storageLoad.png](./Diagrams/Storage/loginStorageSequenceDiagram.drawio.png)

Reviewer: maybe should show diagrams first before explaining the implementation
